### PR TITLE
Refactorings for loading this plugin faster

### PR DESCRIPTION
**Pull request by rhysd**
I did some refactorings mainly for loading this plugin faster (and fixing some global scope pllution).

- Move implementation into `autoload/` directory to defer loading plugin until omni completion is executed
- Fix variable prefixes not to pollute global scope and to stop variables living longer unnecessarily
- Improve file path separator for Windows
- Fix indentation

### CODE REVIEW

File: autoload/racer.vim [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+    let out = []
+
+    for line in lines
+        if line =~ "^MATCH"
+            let completions = split(line[6:], ",")
+            let kind = get(typeMap, completions[4])
+            let completion = {'kind' : kind, 'word' : completions[0], 'dup':1 }
+
+            if kind ==# 'f' " function
+                let completion['menu'] = substitute(substitute(substitute(join(completions[5:], ','), '\(pub\|fn\) ',"","g"), '{*$', "", ""), ' where\s\?.*$', "", "")
```
Comment (Shougo) [inline autoload/racer.vim]: Too Long line.

File: autoload/racer.vim [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+                if g:racer_insert_paren == 1
+                    let completion['abbr'] = completions[0]
+                    let completion['word'] .= "("
+                endif
+                let completion['info'] = join(completions[5:], ',')
+            elseif kind ==# 's' " struct
+                let completion['menu'] = substitute(substitute(join(completions[5:], ','), '\(pub\|struct\) ',"","g"), '{*$', "", "")
```
Comment (Shougo) [inline autoload/racer.vim]: Too Long line.

File: autoload/racer.vim [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+endfunction
+
+function! s:RacerGetBufferContents(base)
+    " Re-combine the completion base word from omnicomplete with the current
+    " line contents. Since the base word gets remove from the buffer before
+    " this function is invoked we have to put it back in to out tmpfile.
+    let col = col(".") - 1
+    let buf_lines = getline(1, '$')
+    let line_contents = getline('.')
+    let buf_lines[line('.') - 1] = strpart(line_contents, 0, col).a:base.strpart(line_contents, col, len(line_contents))
```
Comment (Shougo) [inline autoload/racer.vim]: Too long line

File: autoload/racer.vim [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+endfunction
+
+function! racer#ShowDocumentation()
+    let winview = winsaveview()  " Save the current cursor position
+    " Move to the end of the word for the entire token to search.
+    " Move one char back to avoid moving to the end of the *next* word.
+    execute "normal he"
+    let col = col('.')
+    let b:tmpfname = tempname()
+    call writefile(getline(1, '$'), b:tmpfname)  " Create temporary file with the buffer's current state
```
Comment (Shougo) [inline autoload/racer.vim]: Too long line

File: autoload/racer.vim [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+    call delete(b:tmpfname)
+    return out
+endfunction
+
+function! s:RacerSplitLine(line)
+    let separator = ';'
+    let placeholder = '{PLACEHOLDER}'
+    let line = substitute(a:line, '\\;', placeholder, 'g')
+    let parts = split(line, separator)
+    let docs = substitute(substitute(substitute(substitute(get(parts, 7, ''), '^\"\(.*\)\"$', '\1', ''), '\\\"', '\"', 'g'), '\\''', '''', 'g'), '\\n', '\n', 'g')
```
Comment (Shougo) [inline autoload/racer.vim]: Too long line

File: autoload/racer.vim [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+    let b:tmpfname = tempname()
+    call writefile(getline(1, '$'), b:tmpfname)  " Create temporary file with the buffer's current state
+    let fname = expand("%:p")
+    let cmd = g:racer_cmd." complete-with-snippet ".line(".")." ".col." ".fname." ".b:tmpfname
+    let res = system(cmd)
+    call winrestview(winview)  " Restore de cursor position
+    call delete(b:tmpfname)  " Delete the temporary file
+    let lines = split(res, "\\n")
+    for line in lines
+       if line =~ "^MATCH"
```
Comment (Shougo) [inline autoload/racer.vim]: It should be:

```vim
if line !~ '^MATCH'
  continue
endif
```

File: autoload/racer.vim [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+        \ 'Crate' : 'C', 'Let' : 'v', 'StructField' : 'm',
+        \ 'Impl' : 'i', 'Enum' : 'e', 'EnumVariant' : 'E',
+        \ 'Type' : 't', 'FnArg' : 'v', 'Trait' : 'T'
+        \ }
+
+    let lines = split(res, "\\n")
+    let out = []
+
+    for line in lines
+        if line =~ "^MATCH"
```
Comment (Shougo) [inline autoload/racer.vim]: It should be:

```vim
if line !~ '^MATCH'
  continue
endif
```

File: autoload/racer.vim [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+    let fname = expand("%:p")
+    let cmd = g:racer_cmd." complete-with-snippet ".line(".")." ".col." ".fname." ".b:tmpfname
+    let res = system(cmd)
+    call winrestview(winview)  " Restore de cursor position
+    call delete(b:tmpfname)  " Delete the temporary file
+    let lines = split(res, "\\n")
+    for line in lines
+       if line =~ "^MATCH"
+           let docs = s:RacerSplitLine(line[6:])[7]
+           if len(docs) > 0  " Only open doc buffer if there're docs to show
```
Comment (Shougo) [inline autoload/racer.vim]: It should be:

```vim
if docs == ''
  continue
endif
```

File: autoload/racer.vim [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+               let bn = bufnr("__doc__")
+               if bn > 0
+                   let wi=index(tabpagebuflist(tabpagenr()), bn)
```
Comment (Shougo) [inline autoload/racer.vim]: It should be:

```vim
let wi = index(tabpagebuflist(tabpagenr()), bn)
```

File: autoload/racer.vim [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+    " this function is invoked we have to put it back in to out tmpfile.
+    let col = col(".") - 1
+    let buf_lines = getline(1, '$')
+    let line_contents = getline('.')
+    let buf_lines[line('.') - 1] = strpart(line_contents, 0, col).a:base.strpart(line_contents, col, len(line_contents))
+    return buf_lines
+endfunction
+
+function! s:RacerJumpToLocation(filename, linenum, colnum)
+    if(a:filename != '')
```
Comment (Shougo) [inline autoload/racer.vim]: It should be:

```vim
if a:filename == ''
    return
endif
```

File: autoload/racer.vim [hunk 11]
```diff
@@ -0,0 +1,264 @@
...
+        \ 'Crate' : 'C', 'Let' : 'v', 'StructField' : 'm',
+        \ 'Impl' : 'i', 'Enum' : 'e', 'EnumVariant' : 'E',
+        \ 'Type' : 't', 'FnArg' : 'v', 'Trait' : 'T'
+        \ }
+
+    let lines = split(res, '\n')
+    let out = []
+
+    for line in lines
+        if line !~ '^MATCH'
```
Comment (Shougo) [inline autoload/racer.vim]: It should be `line !~# '^MATCH'`

File: autoload/racer.vim [hunk 12]
```diff
@@ -0,0 +1,264 @@
...
+    let fname = expand('%:p')
+    let cmd = g:racer_cmd . ' complete-with-snippet ' . line('.') . ' ' . col . ' ' . fname . ' ' . b:tmpfname
+    let res = system(cmd)
+    " Restore de cursor position
+    call winrestview(winview)
+    " Delete the temporary file
+    call delete(b:tmpfname)
+    let lines = split(res, '\n')
+    for line in lines
+        if line !~ '^MATCH'
```
Comment (Shougo) [inline autoload/racer.vim]: It should be `line !~# '^MATCH'`

File: autoload/racer.vim [hunk 13]
```diff
@@ -0,0 +1,264 @@
...
+    let fname = expand('%:p')
+    let tmpfname = tempname()
+    call writefile(getline(1, '$'), tmpfname)
+    let cmd = g:racer_cmd . ' find-definition ' . line('.') . ' ' . col . ' ' . fname . ' ' . tmpfname
+    let res = system(cmd)
+    let lines = split(res, '\n')
+    for line in lines
+        if res =~# ' error: ' && line !=# 'END'
+            call s:Warn(line)
+        elseif line =~ '^MATCH'
```
Comment (Shougo) [inline autoload/racer.vim]: It should be `elseif line =~# '^MATCH'`

File: autoload/racer.vim [hunk 14]
```diff
@@ -0,0 +1,264 @@
...
+        silent normal! 1Gdd
+        setlocal nomodifiable
+        setlocal nomodified
+        setlocal filetype=rustdoc
+        break
+    endfor
+endfunction
+
+function! s:RacerGetCompletions(base)
+    let col = col(".") - 1
```
Comment (Shougo) [inline autoload/racer.vim]: `col('.')`

File: autoload/racer.vim [hunk 15]
```diff
@@ -0,0 +1,264 @@
...
+    let b:tmpfname = tempname()
+    " HACK: Special case to offer autocompletion on a string literal
+    if getline(".")[:col-1] =~# '".*"\.$'
```
Comment (Shougo) [inline autoload/racer.vim]: `getline('.')`

**Comment by rhysd** [comment]
I believe I fixed all the points. Could you review additional fixes? Diff is below:

https://github.com/rhysd/vim-racer/compare/d6f3d30803d193406fcbb8f5130eb993137749a9...2b029aab0e97fd32a18085473734ab3cc055f31c

**Comment by Shougo** [comment]
Almost good.
Please fix the new comments.

**Comment by rhysd** [comment]
Thank you for catching them. I fixed.

**Comment by Shougo** [comment]
Nice.  Merged.

**Comment by rhysd** [comment]
Thank you for your review.

**Pull request merged**